src/LoginPage.cc

```diff
@@ -142,63 +177,111 @@ void LoginPage::loginError(QString error)
 	error_label_->setText(error);
 }
 
-void LoginPage::onLoginButtonClicked()
+void LoginPage::onMatrixIdEntered()
 {
+	guessing_server = true;
 	error_label_->setText("");
-
-	if (!matrixid_input_->hasAcceptableInput()) {
-		loginError("Invalid Matrix ID");
-	} else if (password_input_->text().isEmpty()) {
-		loginError("Empty password");
+	QString home_server = matrixid_input_->text().split(":").at(1);
+	server_input_->setText(home_server);
+	client_->setServer(home_server);
+	client_->versions();
+
+	if (server_input_->isVisible()) {
+		server_error_->hide();
+		server_success_->hide();
+		server_spinner_->show();
 	} else {
-		QString user = matrixid_input_->text().split(":").at(0).split("@").at(1);
-		QString password = password_input_->text();
+		matrixid_success_->hide();
+		matrixid_spinner_->show();
+	}
 
-		QString home_server = custom_domain_.isEmpty()
-					      ? matrixid_input_->text().split(":").at(1)
-					      : custom_domain_;
+	server_layout_->invalidate();
+	matrixid_layout_->invalidate();
```

Comment:
The layouts wouldn't realign themselves when hiding things otherwise. That might be different if the hidden elements are actually removed from the layout.